enhancements/testing/build_root_golang_version_from_repo.md

```diff
@@ -0,0 +1,130 @@
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Today, the image in which binaries are built by ci-operator (`build_root_image`) can be
+configured by either specifying an imagestream or by specifying a Dockerfile in the
+repository to build this image from. In the latler case, the Dockerfile is in the basebranch
+of PRs. This means its not possible to do a PR that atomically changes both contents of the
+repository and the `build_root_image`. This proposal proposes a way to enable changing both
+code and the `build_root_image` atomically with one PR.
+
+
+## Motivation
+
+### Goals
+
+* Allow changing both repository contents and `build_root` in a single PR
+
+### Non-Goals
+
+* Change anything about the concept of the `build_root` or how we build images
+  for promotion
+
+## Background
+
+The ci-operator config has a `build_root` section that can either include an imagestreamtag
+or a Dockerfile. If it is an Dockerfile, a build of that Dockerfile from the base branch of the
+PR will be created. In both cases, the resulting image will be tagged into the `pipeline:root`
+imagestreamtag.
+
+The `pipeline:root` imagestreamtag in turn is used to:
+* Build binaries via the `binary_build_command` and tag them into the `pipeline:bin` imagestreamtag
+* Build test binaries via the `test_binary_build_commands` and tag them into the `pipeline:test-bin` imagestreamtag
+* Build rpms via the `rpm_build_commands` and tag them into the `pipeline:rpms` imagestreamtag
+
+All these tags in turn can then be refenced in tests.
+
+Additionally, if the project promotes, there is one Dockerfile per promoted image. It usually duplicates
+the `build_root` and `binary_build_commands` but copies the resulting binaries in an `ocp/$version:base`
+image.
+
+## Proposal
+
+It is proposed to extend the existing `build_root` section in the ci-operator config
+with a new `from_repository` boolean that will indicate of the image should be inferred
+from the repository:
+```
+build_root:
+	from_repository: true
+```
+
+This setting will be mutually exclusive with all other options in the `build_root` section.
+If set, the `ci-operator-prowgen` cli will be changed to clone the repository into the ci
+operator pod. This in turn allows the `ci-operator` to read the image to be used from the
+repository.
+
+The `build_root_image` must not be a raw docker pullspec but an imagestreamtag, because
+we tag it into the `pipeline:root` imagestreamtag and have found that doing an import for
+every job run causes performance issues in the underlying cluster.
+
+TODO: Agree on which of the two options below is better.
+
+There are two options to to get the imagestreamtag from:
+
+### A new configfile in the repository
```

Comment:
does it make sense to make this an extensible yaml file rather than a file with the build_root_image only?